076-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,389 @@
+  If in future components are able to dynamically update their TrustStores, updating the Kubernetes Secret storing the CA certificate doesn't automatically update the CA certificate being trusted.
+  This also means when integrating with a certificate management solution Strimzi doesn't need to reuse these Kubernetes Secrets, it can fetch certificates programmatically for example.
+* It will be easier to reason about which CA certificates are in use/trusted, both as an end user when viewing the cluster, but also as a developer when reviewing the code.
+  This is because currently the state is inferred by observing various Kubernetes annotations and Java class fields.
+
+These benefits will also make it easier to evolve the management of CA certificates in future (see [Future changes](#future-changes)).
+
+Non-goals of the proposal:
+* This proposal does not change the management of Clients CA.
+  Users can provide custom listener certificates, so allowing integration with a certificate management solution for the Clients CA doesn't bring as many benefits as it does for the Cluster CA.
+* This proposal does not introduce integration with a certificate management solution.
+  The aim of this proposal is to make it easier to allow this kind of integration in future, but assumes that the Cluster CA is still managed by Strimzi and (other than during renewal) there is a single Cluster CA certificate for the cluster.
+
+### New trust-state Kubernetes Secret
+
+A new Kubernetes Secret is introduced which is updated by the cluster operator in its role of managing the cluster (specifically which issued certificates are stored where and rolling pods to use/trust those certificates) to reflect whether the individual CA certificates are trusted and/or in use in the cluster.
+
+```yaml
+kind: Secret
+metdata:
+  name: <CLUSTER_NAME>-cluster-ca-trust-state # e.g. foo-cluster-ca-trust-state
+data:
+  ${fingerprint}: ${state}
+```
+
+The `fingerprint` is the fingerprint/thumbprint of the certificate. I.e. the SHA-1 hash of the ASN.1 DER-encoded form of the X509 certificate, which is a commonly used way of identifying certificates using common tooling.
+The `state` can be any of the states of the CA certificate trust state machine described below.
+
+### CA certificate state machine
+
+The "CA certificate trust state machine" has the following states:
+
+```
+  UNTRUSTED // not yet trusted everywhere it needs to be 
+   |
+   v
+  TRUSTED_UNUSED // trusted everywhere it needs to be, but no EE certificates yet issued with this CA certificate
+   |
+   v
+  TRUSTED_IN_USE // trusted everywhere, EE certificates issued using this CA certificate
+   |
+   v
+  PHASE_OUT // trusted everywhere, but new end-entity certificates will not be issued using this CA certificate
+   |
+   v
+  ZOMBIE // still trusted by some component, but with no issued end-entity certificates in current use.
+```
```

Comment:
How do you expect to manage this if as you suggest in the motivation "you do not know what CA will be used by the next certificate"? You will never know that something is t be phased out or zombie.

Reply:
This proposal doesn't fully address that problem, it focuses on better tracking the CA certificates within the cluster and would allow us to later have multiple CAs in use (beyond the current one and a single previous one as we do today). The missing part I think you are eluding to here is if we changed to have cert-manager issuing certificates we would need some way to track/understand which CA certificates are associated with which end-entity certificates, but I didn't include those details here as I was trying to keep the proposal more narrowly scoped. We need a separate proposal that looks at changes to how we request and issue certificates internally.

Reply:
To be honest, I do not think this adds up:
* You include some motivation for this proposal
* You introduce a set of states that clearly do not support the motivation
* And then you say that this is a missing part and will be addressed later in another proposal

I think this will be a major effort to implement in a way that delivers some value. So we have to make sure it is really preparing for the future.

Reply:
@scholzj is correct that the transition `TRUSTED_IN_USE -> PHASE_OUT` cannot be one way, because we don't know whether a new EE cert will, in the future, be issued using this CA cert. We should describe the conditions necessary for a transition to occur, rather than just the states. In this case I wonder if we really need both `PHASE_OUT` and `ZOMBIE`. 

Let's consider a particular CA cert called $p$ which has been used to sign a set of EE certificates $X$ which are each being used some a component.

1. While $|X| > 0$, $p$ is in state `TRUSTED_IN_USE`.
2. We request a replacement cert for $x \in X$, the last EE cert in $X$ that was signed by $p$. Its replacement, $x'$ is signed by a different CA cert $p'$. $p$ is in state `TRUSTED_IN_USE`, $p'$ is in state `UNTRUSTED`. $x$ is still being used. We can't use $x'$ yet because $p'$ is not trusted.
3. Because $p'$ is not in state `TRUSTED_UNUSED` or `TRUSTED_UNUSED` and is the signer of some EE cert ($x'$) we roll components. 
4. Once everything has been restarted (which we can tell from the annotations) we know everything trusts the new $p'$, so we transition $p'$ to `TRUSTED_UNUSED`.
5. We can now replace $x$ with $x'$ in the secret consumed by the thing that was using $x$, and restart that component.
6. Once we know that $x'$ is in use (which we see from the annotation) we know that $x$ isn't in use any more. So we transition $p'$ to `TRUSTED_USED`.
7. We see that nothing in $X$ was signed by $p$, so we transition $p$ to `PHASE_OUT`.
8. Because $p$ is in `PHASE_OUT` we start rolling components which trust $p$.
9. Once the last such component has restarted, we find (from the annotations) that nothing actually trusts $p$ any more. If $p$ is still in `PHASE_OUT` we forget about $p$.

The point @scholzj is making is that between 8 and 9 it's possible we end up requesting some other EE cert (for another component, say), which happens to be signed by $p$. This is why we'd need the last sentence in step 9, and why in step 3 the condition is not "$p'$ is in state `UNTRUSTED`" and also why we have the requirement that it $p'$ is the signer of _something_.

The conditions used to drive a transition are always taken from querying resources in Kube (not based on in-memory knowledge of what the operator has just done). This should mean that the operator can crash/restart at any point, and that the above transitions can happen across multiple reconciliations, but still arrive at the same outcome.